subgraphs/euler-finance/src/mappings/helpers.ts

```diff
@@ -794,6 +810,137 @@ export function processReward(epoch: _Epoch, epochStartBlock: BigInt, event: eth
+          epoch.id,
+        ]);
+      }
+
+      // reset mkt._weightedStakedAmount for the new epoch
+      mkt._weightedStakedAmount = BIGINT_ZERO;
+      // EUL staked remains staked for the market until unstaked
+      // so not reset mkt._stakedAmount
+      mkt.save();
+    }
+  }
+}
+
+export function processRewardEpoch18_23(epoch: _Epoch, epochStartBlock: BigInt, event: ethereum.Event): void {
+  // eIP24 changes the reward distribution
+  // https://snapshot.org/#/eulerdao.eth/proposal/0x7e65ffa930507d9116ebc83663000ade6ff93fc452f437a3e95d755ccc324f93
+  const epochID = epoch.epoch;
+  const prevEpochID = epochID - 1;
+  const prevEpoch = _Epoch.load(prevEpochID.toString());
+  if (prevEpoch) {
+    const protocol = getOrCreateLendingProtocol();
+    // finalize mkt._weightedStakedAmount for prev epoch & distribute rewards
+    // The array is needed to select top 10 staked markets
+    // const marketWeightedStakedAmounts: BigInt[] = [];
+    let sumAccumulator = BIGDECIMAL_ZERO;
+    for (let i = 0; i < protocol._marketIDs!.length; i++) {
+      const mktID = protocol._marketIDs![i];
+      const mkt = Market.load(mktID);
+      if (!mkt) {
+        log.error("[handleStake]market {} doesn't exist, but this should not happen at tx ={}", [
+          mktID,
+          event.transaction.hash.toHexString(),
+        ]);
+        continue;
+      }
+
+      // eIP28 blacklist FTT market from EUL distribution
+      // https://snapshot.org/#/eulerdao.eth/proposal/0x40874e40bc18ff33a9504a770d5aadfa4ea8241a64bf24a36777cb5acc3b59a7
+      if (epochID >= 16 && mkt.inputToken == FTT_ADDRESS) {
+        mkt._stakedAmount = BIGINT_ZERO;
+        mkt._weightedStakedAmount = BIGINT_ZERO;
+      }
+
+      const stakedAmount = mkt._stakedAmount;
+      let mktWeightedStakedAmount = BIGINT_ZERO;
+      if (isMarketEligible(mkt) && stakedAmount.gt(BIGINT_ZERO)) {
+        // finalized mkt._weightedStakedAmount for the epoch just ended
+        // epochStartBlock.minus(BIGINT_ONE) is the end block of prev epoch
+        updateWeightedStakedAmount(mkt, epochStartBlock.minus(BIGINT_ONE));
+        mktWeightedStakedAmount = mkt._weightedStakedAmount!;
+      }
+      mkt.save();
+
+      sumAccumulator = sumAccumulator.plus(mktWeightedStakedAmount.sqrt().toBigDecimal());
+    }
+
+    const EULPriceUSD = getEULPriceUSD(event);
+    const borrowerRewardToken = getOrCreateRewardToken(Address.fromString(EUL_ADDRESS), RewardTokenType.BORROW);
+    const lenderRewardToken = getOrCreateRewardToken(Address.fromString(EUL_ADDRESS), RewardTokenType.DEPOSIT);
+    const totalRewardAmount = BigDecimal.fromString((EUL_DIST[epochID - START_EPOCH] * EUL_DECIMALS).toString());
+
+    // scale to daily emission amount
+    const dailyScaler = BigDecimal.fromString((BLOCKS_PER_DAY / (BLOCKS_PER_EPOCH as f64)).toString());
+    const EUL_DECIMALS_BD = BigDecimal.fromString(EUL_DECIMALS.toString());
+    for (let i = 0; i < protocol._marketIDs!.length; i++) {
```

Comment:
This is the same for loop as the one above.

It opens the door to bugs down the road since you now have 2 places doing the same thing.

Can we make this all 1 for loop?

Reply:
Also, this is a pretty huge for loop. Is it possible to move anything to a separate function?

Reply:
I know. IRCC, we went through this before, I don't think I can combine the two loops: [the first loop](https://github.com/tnkrxyz/subgraphs/blob/f738023468d9f15cc7f1a71cc49f3493583afee7/subgraphs/euler-finance/src/mappings/helpers.ts#L843-L872) sums the weighted staked amount, and then [the second loop](https://github.com/tnkrxyz/subgraphs/blob/f738023468d9f15cc7f1a71cc49f3493583afee7/subgraphs/euler-finance/src/mappings/helpers.ts#L882-L951) allocates EUL among eligible markets proportional to the sum weighted staked amount (calculated from the first loop). Maybe I can collect the eligible markets into one array and _weightedStakedAmount into another, but I will still have to loop through them - calculating the daily reward amount, and add emission amount for the few singled out markets - pretty not helping much.

Reply:
No worries, we can leave for now. And yes I remember having this convo lmao. I am resolving